### ISSUE TITLE
New LLE render.

### ISSUE THREAD

**Issue by gonetz** [gonetz/GLideN64#2169]
The goal is to fix numerous issues with LLE rendering. Ultimate goal is to make LLE as good as HLE.

I implemented new method to map N64 LLE triangle into set of OpenGL triangles.
The idea and implementation details described in my articles:
http://gliden64.blogspot.com/2019/10/rendering-in-low-level-emulation-mode.html
http://gliden64.blogspot.com/2019/11/rendering-in-low-level-emulation-mode.html
http://gliden64.blogspot.com/2019/11/rendering-in-low-level-emulation-mode_17.html

I still can't solve the problem with depth fighting in decal mode, so the code is in dev branch https://github.com/gonetz/GLideN64/tree/new_lle. I pushed it as one commit, without dozen of intermediate ones. I made some code cleanup, but it still contains plenty of experimental stuff. 

Related issues:
* Killer Instinct Gold Skies Aren't Displaying Correctly #406
* Goldeneye Texture errors and slowdowns #459
* Corrupt Map Lines. (Turok, Doom, Winback.) #465
* World Driver Championship Texture Clipping\Corruption. #466
* Indiana Jones and the Infernal Machine Missing Sprites in LLE #467
* Duke Nukem 64 Missing Sprites. #468
* Yoshis Story - Black Background and Yoshi Duplicates with LLE #553
* depth buffer emulation in LLE #671
* Pokemon Stadium 2 validity check fails in LLE in the library #725
* Glover missing sprites #842
* vivid dolls (aleck64): incorrect gfx #850
* F-Zero X,  #1333
* Weird brightness striping issue #1401
* SD Hiryuu no Ken Densetsu: OpenGL Error in LLE #1736
* Pokemon Snap LLE Render 2D Elements in N64 Res Error #1945 

Test build:
https://drive.google.com/file/d/1CsByo95B9UC2uCAZ3LUY02mANoU1VQBp/view?usp=sharing

The new build should work as good or better than the current master.
There should be no regressions in LLE mode.
If you will notice some regressions, please report asap.
Also please check issues from "Related issues" list. I'm sure that some of them are fixed with the new render.

**Comment by DonelBueno** [comment]
The only I could check was Pokémon Snap in LLE (with Render 2D Elements in N64) and couldn't find any issue.

**Comment by AriaHiro64** [comment]
In dragon sword (an old unreleased game) this fixed the sparks (the master shows weird lines)

**Comment by gonetz** [comment]
I didn't know about that issue in dragon sword. Old LLE can't render lines correctly.

**Comment by fzurita** [comment]
It works fine in Android. Killer instinct skies are fixed! Pure LLE also works. It's seems to be around 5 to 10% faster overall. The CXD4 RSP hold it back I think.

**Comment by gonetz** [comment]
> It works fine in Android.

Great, thanks!

@IronYoshi64 Could you please check, which issues from the "Related issues" list are fixed with this version? I'm going to merge this feature branch to master and I need to know, which issues I can close after that.

**Comment by purplemarshmallow** [comment]
@DonelBueno I tested https://github.com/gonetz/GLideN64/issues/725 with an older and newer build and it seems fixed now

**Comment by purplemarshmallow** [comment]
The snow in Ploaris SnoCross is a special case for the triangle function. z64gl has buggy snow because of that. GLideN64 has missing snow in LLE now and with older builds.

**Comment by gonetz** [comment]
@purplemarshmallow thanks for the report about Ploaris SnoCross. I wasn't aware about this issue.
Show rendered with gSPLineW3D command. It is a line, nothing special. It was not rendered with new LLE code because of depth compare issue. Triangle command for snow does not use depth, but depth compare is enabled in othermode. I fixed that issue in 0b492bf3ca

**Comment by fzurita** [comment]
@gonetz is it worth merging this as is? I don't see what could break since this only affects LLE, and I don't think the current LLE implementation could be made worse by the new implementation.

HLE games that use some LLE command seem only improved from this.

**Comment by gonetz** [comment]
It's kinda incomplete and unpolished. I'll try to make some corrections and then merge it.

Btw, @fzurita, I need to ask you as an expert in mobile CPUs. I use mostly double floats in my new LLE. PC hardware has no problems with doubles, but I'm not sure about mobiles. Could you share your opinion about it?
I'm also not sure that doubles are really necessary here. It was an experimental code and I did not want any problems with precision. Probably 32bit float will be fine as well.

**Comment by fzurita** [comment]
I wouldn't call myself an expert in mobile CPUs, even though I do deal with them a lot. From a bit of research though, if doubles are used then SIMD operations (NEON, SSE, AVX) will not be as efficient. So if the compiler does apply any SIMD optimizations, they will not be as effective. This is true for both x86 and ARM.

Also, from a bit of reading double precision is not supported at all for the NEON instruction set. So if doubles are used, the compiler would not be able to automatically vectorize code, from here http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dht0004a/ch01s04s03.html:

> Also, NEON technology does not support some data types, and some are only supported for certain operations. For example, it does not support double-precision floating-point, so using a double-precision double where a single-precision float is sufficient can prevent the compiler from vectorizing code. NEON technology supports 64-bit integers only for certain operations, so avoid using long long variables where possible.


If no SIMD operations are used, double will still have a performance impact on some ARM chipsets, specially if they don't have an FPU, not that I would recommend running running GLideN64 on a chipset without an FPU though.

Here are some benchmarks I quickly dug up:
http://nicolas.limare.net/pro/notes/2014/12/12_arit_speed/

**Comment by gizmo98** [comment]
At least ARMv8 NEON supports doubles and is IEEE754 compliant. ARMv7 auto vectorize only works if -ffast-math and -ftree-vectorize is enabled.  

According to this post double perfomance is 1/2 or 1/4 of float performance:
https://community.arm.com/developer/ip-products/processors/f/cortex-a-forum/6552/why-does-fpu-performance-differ-in-aarch64-and-aarch32-with-cortex-a53

**Comment by gonetz** [comment]
Thanks. So, at least it worth to try to switch to 32bit floats.

**Comment by olivieryuyu** [comment]
2 issue

Depth issue in Duke Nukem 64

it is not decals in my opinion

![GLideN64_DUKE_NUKEM_000](https://user-images.githubusercontent.com/8987490/75587679-4d237980-5a77-11ea-857b-87c0137cec4c.png)

wrong textures in intro

It is not decals as well in my opinion

![GLideN64_Command Conquer_000](https://user-images.githubusercontent.com/8987490/75587776-878d1680-5a77-11ea-8374-c85e9d289c00.png)

**Comment by gonetz** [comment]
> Depth issue in Duke Nukem 64
> 
> it is not decals in my opinion

Yes, it is not a problem with decals. LLE code calculates depth a bit incorrect. Sprites are texrtects. Depth for sprites provided by primitive_depth. If depth of 3D geometry is incorrect, 2D sprites are either invisible because they are behind 3D polygons (as with old LLE) or they become visible too early. The problem is that I don't know why the depth is incorrect and how to fix it.

**Comment by gonetz** [comment]
> wrong textures in intro

Fixed with commit 4ead76e207

**Comment by weinerschnitzel** [comment]
Could the Duke nukem depth issue be stepped through with angrylion to analyze how the scene is rendered to better understand the depth issue and what is wrong?

My understanding with the gaps between polygons issue was that edge walking was not accurate enough. Would a branch for accurate edge walking be a good move in the right direction? Maybe at least to get a proof of concept if that approach has benefit? The issue is also seen in HLE, so I am a bit unsure if LLE work would help here. Finding intersections sounds like the ideal approach for the plugin, though.

**Comment by gonetz** [comment]
> Could the Duke nukem depth issue be stepped through with angrylion to analyze how the scene is rendered to better understand the depth issue and what is wrong?

Yes, sure. Software plugin works with the same data as hardware one, so it is possible to track how the data is processed. It is hard and tedious task.


> My understanding with the gaps between polygons issue was that edge walking was not accurate enough. Would a branch for accurate edge walking be a good move in the right direction?

I have no idea how to implement accurate edge walking for hardware rendering.

**Comment by olivieryuyu** [comment]
Supercross 2000 in LLE

![GLideN64_Supercross_000](https://user-images.githubusercontent.com/8987490/76171345-6076c900-618a-11ea-8fd8-60a40fe18e09.png)

May be fixed by 4ead76e?

**Comment by mudlord** [comment]
> I have no idea how to implement accurate edge walking for hardware rendering.

That means the edge/slope traversal > LLE GL triangle conversion, which you are now working on.

Wouldn't it make sense for quads though to render as triangle strips/fans? That way you don't have to worry about doing two triangles each quad.

**Comment by gonetz** [comment]
> That means the edge/slope traversal > LLE GL triangle conversion, which you are now working on. And you said a while back now, ever since working on the new LLE render code, it didn't seem to make a difference.

As I understand, accurate edge walking means accurate reproduction of RDP rendering mechanism. I don't think that it is possible with just LLE GL triangle conversion, no matter how good and precise it is.  RDP does some specific logic on edges of polygons, which is hardly can be reproduced with PC hardware. The only way I see is to render LLE triangle not with draw triangle command but as set of lines, spam by span as RDP does. BTW, do you know how polygons rendering implemented in Vulcan port of AL RDP?



> Wouldn't it make sense for quads though to render as triangle strips/fans? That way you don't have to worry about doing two triangles each quad.

What is the benefit? Performance? As I see, as soon as you have some triangles caching implemented, it does not matter how they are represented, as strips or as independent triangles. Numbers of API calls matters, not number of triangles to draw with each call. I implemented triangles caching for LLE and got huge performance boost.

**Comment by gonetz** [comment]
> May be fixed by [4ead76e](https://github.com/gonetz/GLideN64/commit/4ead76e207e21965ee99379e33c3aea2c1400ae8)?

Fixed, but not completely. It now looks as with master:
![GLideN64_Supercross_000](https://user-images.githubusercontent.com/8983092/76194156-29e78f80-6218-11ea-8014-568bcc8a0077.jpg)

**Comment by mudlord** [comment]
> BTW, do you know how polygons rendering implemented in Vulcan port of AL RDP?

Its implemented as compute shaders. Everything is done in compute shaders. Even span drawing.

>As I see, as soon as you have some triangles caching implemented, it does not matter how they are represented, as strips or as independent triangles. Numbers of API calls matters, not number of triangles to draw with each call. I implemented triangles caching for LLE and got huge performance boost.

I assume you are handling duplicate vertices when drawing quads with triangles too, via index buffers. 

In that case if you are buffering already, yeah, you have a point. Shouldn't make much difference. Just thought using triangle strips or fans might simplify code a bit.

**Comment by gonetz** [comment]
> Its implemented as compute shaders. Everything is done in compute shaders. Even span drawing.

Thanks! It's smart. Shame to me, I did not learn about compute shaders until now. I've got the idea: angrylion's code ported to compute shaders, which compute color and depth and store them into output textures. I suppose these textures are synced with RDRAM.
Yes, this way we can get hardware rendering without use of hardware's drawing capabilities.
However, I doubt that it will work properly for non-native resolutions.

**Comment by gonetz** [comment]
> Fixed, but not completely.

Completely fixed with af754f05b

**Comment by mudlord** [comment]
>Yes, this way we can get hardware rendering without use of hardware's drawing capabilities.

Yes, Vulkan opens the opportunities for many things that OpenGL just cannot do. Also a key feature of Vulkan was asynchronous compute. Some games like the recent Doom game use this, as well as the Vulkan Angrylion port/rewrite.

https://gpuopen.com/concurrent-execution-asynchronous-queues/

AFAIK the main problem with the Vulkan Angrylion port is HWFBE support (its very incomplete). I guess that comes down to how Themaister understood HWFBE and how some games use the framebuffer.

**Comment by gonetz** [comment]
> AFAIK the main problem with the Vulkan Angrylion port is HWFBE support (its very incomplete). I guess that comes down to how Themaister understood HWFBE and how some games use the framebuffer.

As soon as you have RDRAM and color/depth buffers separated, HWFBE problem becomes inevitable.


> OGL3/4 has things like geometry shaders, but I don't have much experience with those, and dunno how those might work for N64 RDP things.

OGL has compute shaders support since 4.3 
There was an idea to  combine Angrylion's RDP with GLideN64, #824
It probably could be done with that Vulcan port.

**Comment by loganmc10** [comment]
@gonetz I just got around to testing this (Intel GPU). I get the following GPU shader compilation 
```
major shader compiler error 1: SHADER_ID_LINK error has been generated. GLSL link failed for program 9, "": Type mismatch: Type of vShadeColor different between shaders.
Out of resource error.
```

**Comment by gonetz** [comment]
I'll check on Intel, thanks!

**Comment by gonetz** [comment]
Shader compilation on Intel GPU is fixed with 2f1cda16541

**Comment by loganmc10** [comment]
Yes looks good on Intel now!

**Comment by AriaHiro64** [comment]
not sure if this is a old or new lle issue but mm's framebuffer is clamped vertically in lle

**Comment by StenApp** [comment]
SD Hiryuu no Ken Densetsu: OpenGL Error in LLE #1736 **and**
Goldeneye regression #1820
are fixed, aren't they?

**Comment by gonetz** [comment]
@StenApp yes, both are closed.

**Comment by ytrezq** [comment]
Hi,
I know this is barely related and additionally I might say mistake. Would things like Vulkan or d3d12 be usefull in terms of performances in that case?
I know any computer is powerfull enough to run games at their original framerate these days, but there are some N64 games which supports very high refresh rates. I m meaning sometimes above 60Hz (using a compatible screen of course).

And in that case reaching those framerates is still difficult without high end hardware.

**Comment by olivieryuyu** [comment]
@gonetz 

I have an idea of the cause of the remaining issues in LLE, mainly decals, missing LLE sprite and coronas.

I believe that I mentioned to you before that GE and PD fixed their decals via the extra parameter in gsptexture (xparam in gbi.h). This parameters is used to clamp dzdy if near zero for decal mode (dixit source code of the microcode).

I believe that the issue is that ZBuffer coefficients used by the triangle RDP commands (DzDx, DzDe, DzDy) are not accurate as you would expect.

If you read the N64 programming manual you can read:

15.5.10 Z Accuracy

Figure 15-14 shows the worst-case error ratio in Z relative to the near and far planes.

Figure 15-14 Z Worst-Case Error

![f15-14](https://github.com/gonetz/GLideN64/assets/8987490/6b357d89-11ce-47b0-ac5b-937359b14bbd)

I believe that the way LLE triangle is managed in GlideN64 does not take this inaccuracies in account.

For decals in PD and GE, as the RDP inaccuracies is corrected by a GBI command, in HLE you have no issue as it will match with you would expect for a decal texture with accurate information. 

But you get issues in LLE because not all ZBuffer coefficients are not taken in account for the rendering of the triangle by GlideN64.

I assume that it is the same with the coronas and the sprite, where Z coefficients (being non accurate) are considered by GlideN64.

Not sure if i am clear enough but I guess you may see what i mean.

**Comment by gonetz** [comment]
> This parameters is used to clamp dzdy of the RDP triangle if near zero for decal mode (dixit source code of the microcode).

LLE graphics plugin emulates RDP only. If ucode command has parameters, which correct RDP triangle input data, such as DzDx, DzDe, DzDy, these parameters are applied by RSP plugin, which builds display list for RDP. RDP works the same with all ucodes. It's emulation is independent of RSP commands.
Unfortunately, emulation of RDP triangle with graphics API like OpenGL or Vulcan is very imprecise by its nature, because RDP triangle is not triangle at all. GLideN64 approximates RDP triangle by building a set of OpenGL ones. Inaccuracy is everywhere.

Regarding Z Accuracy. IIRC, the problem is that we should store 18bit of pixel's Z into 16bit depth buffer. RDP uses non-liner transformation there. Z values close to Z-near plane translated 1 to 1, but starting from 128*Znear it becomes more dense, that is close but different Z values can get the same value in the depth buffer. Usually the plugin does not need to emulate that inaccuracy, but for RE2 it is necessary, because the game uses pre-rendered depth buffer, and z values of 3D objects should correspond that buffer.

I know that computation of Z for vertices of LLE triangle isn't good. "But you get issues in LLE because not all ZBuffer coefficients are taken in account for the rendering of the triangle by GlideN64" - there are not so many of them. The formula is smth like
vtx.z = z + dzde * diffY + dzdx * diffX
that is initial Z plus delta Z multiplied by X and Y offsets. Similar formulas are for color components and texture coordinates.
All calculations are in floats, so various side effects related to integer overflow are ignored. As I said, it is an approximation. I suppose the problems with LLE Z inaccuracy can be solved, but it is so tedious task.

**Comment by olivieryuyu** [comment]
I see what you mean, though few points come to my mind:

1. when gliden64 runs in LLE, it actually calculates the vertex information from the RDP information which are reprocessed in OpenGL. As such it is like HLEing the pipeline (sort of). When you calculate vtx.z you consider that the pipeline (meaning the way RSP transform the vertex to RDP command) is fixed, taking the vertex data without modification. However, due to an issue in the RDP itself, decals mode does not work correctly so (and thus only for GE and PD), the pipeline can be influenced by parameters of the ucode. So when calculating vtx.z, you have to take in account this fact in my opinion.
2. I have an idea: if i reverse engine the pipeline RSP code, would it help to get accurate vtx.z? Would it help?

**Comment by gonetz** [comment]
RDP has fixed functionality, it's true. RSP is controlled by ucode, RSP is not.
RDP works the same for all ucodes. "devs of GE and PD modified the pipeline RSP code to amend the issue" - LLE render can't take that fact into account, because that fact exists only on RSP level. RDP emulator does not know about it.
So no, if you reverse engine the pipeline RSP code, it wouldn't help to get better emulation.

Again, RSP prepares display list for RDP, including data for RDP triangle commands.
Devs of GE and PD noticed that the ucode they were working with had issues with decals.
They found that the problem was in the RDP itself. They couldn't fix RDP, it is unchangeable. So, they fixed ucode command, which generated RDP triangle data for decals. That is they changed parameters, which RDP uses to calculate pixel's Z.
It is RSP plugin task to process that modified command and generate correct triangle data.
RDP plugin just should render the triangle using that data.

**Comment by olivieryuyu** [comment]
ok, let me know if i can be of any help, always happy to support
